include/exadg/operators/variable_coefficients.h

```diff
@@ -23,168 +23,213 @@
-                                coefficient_type const & value)
+                                coefficient_type const & coefficient)
   {
-    coefficients_face_neighbor[face][q] = value;
+    coefficients_face_neighbor[face][q] = coefficient;
   }
 
-  // TODO
-  //
-  //  coefficient_type
-  //  get_coefficient_cell_based(unsigned int const face,
-  //                             unsigned int const q) const
-  //  {
-  //    return coefficients_face_cell_based[face][q];
-  //  }
-  //
-  //  void
-  //  set_coefficient_cell_based(unsigned int const       face,
-  //                             unsigned int const       q,
-  //                             coefficient_type const & value)
-  //  {
-  //    coefficients_face_cell_based[face][q] = value;
-  //  }
+  /**
+   * Returns the coefficient in the entry (@p cell_based_face , @p q) in the cell-based face
+   * coefficient table.
+   */
+  coefficient_type
+  get_coefficient_face_cell_based(unsigned int const cell_based_face, unsigned int const q) const
+  {
+    return coefficients_face_cell_based[cell_based_face][q];
+  }
+
+  /**
+   * Sets the specific entry (@p cell_based_face , @p q) in the cell-based face coefficient table
+   * to @p coefficient.
+   */
+  void
+  set_coefficient_face_cell_based(unsigned int const       cell_based_face,
+                                  unsigned int const       q,
+                                  coefficient_type const & coefficient)
+  {
+    coefficients_face_cell_based[cell_based_face][q] = coefficient;
+  }
 
 private:
+  /**
+   * Reinitializes the tables in use. This resizes the tables and initializes the coefficients with
+   * the default constructed @p coefficient_type.
+   */
   template<int dim, typename Number>
   void
   reinit(dealii::MatrixFree<dim, Number> const & matrix_free, unsigned int const quad_index)
   {
     coefficients_cell.reinit(matrix_free.n_cell_batches(), matrix_free.get_n_q_points(quad_index));
 
-    coefficients_face.reinit(matrix_free.n_inner_face_batches() +
-                               matrix_free.n_boundary_face_batches(),
-                             matrix_free.get_n_q_points_face(quad_index));
-
-    coefficients_face_neighbor.reinit(matrix_free.n_inner_face_batches(),
-                                      matrix_free.get_n_q_points_face(quad_index));
-
-    // TODO
-    // // cell-based face loops
-    // coefficients_face_cell_based.reinit(matrix_free.n_cell_batches()*2*dim,
-    // matrix_free.get_n_q_points_face(quad_index));
+    if(use_faces)
+    {
+      coefficients_face.reinit(matrix_free.n_inner_face_batches() +
+                                 matrix_free.n_boundary_face_batches(),
+                               matrix_free.get_n_q_points_face(quad_index));
+      coefficients_face_neighbor.reinit(matrix_free.n_inner_face_batches(),
+                                        matrix_free.get_n_q_points_face(quad_index));
+
+      if(use_cell_based_faces)
```

Comment:
@btemuer Do you think it makes sense to add a comment here why we need all coefficient tables in case of cell-based face loops (I mean also the `coefficients_face` and `coefficients_face_neighbor` tables above)?

Reply:
I don't think it would be better to have that explanation here. The variable coefficients class is better encapsulated, when it just "does what it's told" without the need to know the reasoning behind. Such an explanation would be much better off in a place where this parameter is defined and explained. This will also be much clearer, when we refactor our parameter interface with nested parameters... Then this boolean will for example occur inside the preconditioner parameters.

However, I agree the name `use_cell_based_loops` somehow implies here that it is the only one to be used. Also respecting your comment from below, I will change the variable names to `store_*`.